x/xcvm/keeper/transfer_intent.go

```diff
@@ -0,0 +1,448 @@
+	}
+	return nil
+}
+
+// verifyReceiptProof verifies that the receipt is included in the block header's receipt trie
+func verifyReceiptProof(blockHeader gethtypes.Header, txReceiptHash []byte, receiptProofBz []byte) error {
+	var receiptProof types.ReceiptProof
+	if err := receiptProof.Unmarshal(receiptProofBz); err != nil {
+		return fmt.Errorf("unmarshal receipt proof: %v", err)
+	}
+
+	receiptsRoot := blockHeader.ReceiptHash
+	if _, err := trie.VerifyProof(receiptsRoot, txReceiptHash, receiptProof); err != nil {
+		return fmt.Errorf("verify proof: %v", err)
+	}
+
+	return nil
+}
+
+// getTxReceiptHash returns the hash of the transaction receipt
+func getTxReceiptHash(txReceipt gethtypes.Receipt) ([]byte, error) {
+	//Get binary representation of txReceipt rlp encoding
+	txReceiptBz, err := txReceipt.MarshalBinary()
+	if err != nil {
+		return nil, fmt.Errorf("get tx receipt hash: %v", err)
+	}
+	txReceiptHash := crypto.Keccak256(txReceiptBz)
+	return txReceiptHash, nil
+}
+
+// verifyReceiptUniqueness verifies that the receipt has not been processed before
+func verifyReceiptUniqueness(store store.KVStore, txReceiptHash []byte, blockHash common.Hash) error {
+	receiptKey := types.GetUsedReceiptKey(txReceiptHash, blockHash)
+	if store.Has(receiptKey) {
+		return types.ErrReceiptAlreadyProcessed
+	}
+
+	store.Set(receiptKey, []byte{0x01})
+	return nil
+}
+
+// verifyPreviousBeaconBlockHeaders verifies that the provided previous beacon block headers are valid and in the correct order
+func verifyPreviousBeaconBlockHeaders(beaconBlockHeaders []*types.BeaconBlockHeader) error {
+	intentBeaconBlockHeader := *beaconBlockHeaders[0]
+	headerHash, err := intentBeaconBlockHeader.Hash()
+	if err != nil {
+		return fmt.Errorf("hash intent beacon block header: %v", err)
+	}
+
+	var currentParentRoot [32]byte
+	for _, header := range beaconBlockHeaders[1:] {
+		copy(currentParentRoot[:], header.GetParentRoot())
+		if currentParentRoot != headerHash {
+			return types.ErrInvalidBlockHeaders
+		}
+		headerHash, err = header.Hash()
+		if err != nil {
+			return fmt.Errorf("hash beacon block header: %v", err)
+		}
+	}
+
+	return nil
+}
+
+// verifyBeaconBlockBody verifies that the provided beacon block body matches the beacon block header's body root and the transaction receipt's block hash
+func verifyBeaconBlockBody(beaconBlockHeader types.BeaconBlockHeader, beaconBlockBodySSZ []byte, txReceipt gethtypes.Receipt) error {
+	var beaconBlockBodyRoot [32]byte
+	beaconBlockBodyRootSlice := beaconBlockHeader.GetBodyRoot()
+	copy(beaconBlockBodyRoot[:], beaconBlockBodyRootSlice)
+
+	var beaconBlockBody prysmtypes.BeaconBlockBody
+	if err := beaconBlockBody.UnmarshalSSZ(beaconBlockBodySSZ); err != nil {
+		return fmt.Errorf("unmarshal beacon block body: %v", err)
+	}
+
+	beaconBlockBodyHash, err := beaconBlockBody.HashTreeRoot()
+	if err != nil {
+		return fmt.Errorf("hash beacon block body: %v", err)
+	}
+	if beaconBlockBodyHash != beaconBlockBodyRoot {
```

Comment:
is this line covered by tests?

Reply:
The happy path case is covered by tests for verifiying that the beacon block body root from the header and the provided beacon block body hash match up, however I don't have a test for the error case here. Do you think additional test coverage for the error case is worth adding?